pkg/vacuum/vacuum.go

```diff
@@ -196,70 +219,148 @@ func (e *Engine) Run(ctx context.Context) {
+	} else if workers < 1 {
+		workers = 1
+	}
+	return int(workers)
+}
+
+// calcNumWorkers calculates the number of workers to use to vacuum chunks. The closer the oldest chunk
+// age gets to autovacuumFreezeMaxAge, the more workers we will use
+func (e *Engine) calcNumWorkers(ctx context.Context, con *pgxpool.Conn, chunks []*chunk) (int, error) {
+	vacuumFreezeMinAge, autovacuumFreezeMaxAge, err := e.getVacuumSettings(ctx, con)
+	if err != nil {
+		log.Error("msg", "failed to get vacuum settings", "error", err)
+		return 0, err
+	}
+	var maxChunkAge int64 = 0
+	for _, c := range chunks {
+		if c.age > maxChunkAge {
+			maxChunkAge = c.age
+		}
+	}
+	w := interpolate(float64(vacuumFreezeMinAge), float64(autovacuumFreezeMaxAge), float64(maxChunkAge), float64(e.parallelism))
+	if len(chunks) < w {
+		// don't spin up more workers than we could possibly use
+		w = len(chunks)
 	}
-	return tables, nil
+	return w, nil
 }
 
 // runWorkers kicks off a number of goroutines to work on the chunks in parallel
 // blocks until the workers complete
-func runWorkers(ctx context.Context, parallelism int, chunks []string, worker func(context.Context, int, <-chan string)) {
-	todo := make(chan string, len(chunks))
+func runWorkers(ctx context.Context, numWorkers int, chunks []*chunk, worker func(context.Context, int, <-chan *chunk)) {
+	todo := make(chan *chunk, len(chunks))
 	var wg sync.WaitGroup
-	wg.Add(parallelism)
-	for id := 0; id < parallelism; id++ {
-		go func(ctx context.Context, id int, todo <-chan string) {
+	wg.Add(numWorkers)
+	for id := 0; id < numWorkers; id++ {
+		go func(ctx context.Context, id int, todo <-chan *chunk) {
 			defer wg.Done()
 			worker(ctx, id, todo)
 		}(ctx, id, todo)
 	}
-	for _, chunk := range chunks {
-		todo <- chunk
+	for _, c := range chunks {
+		todo <- c
 	}
 	close(todo)
 	wg.Wait()
 }
 
+// getAutovacuumCount gets the current number of times autovacuum has vacuumed a table
+func (e *Engine) getAutovacuumCount(ctx context.Context, con *pgxpool.Conn, name string) (autovacuumCount int64, err error) {
+	err = con.QueryRow(ctx, sqlGetAutovacuumCount, name).Scan(&autovacuumCount)
+	if err != nil {
+		return 0, err
+	}
+	return
+}
+
 // worker pulls chunks from a channel and vacuums them
-func (e *Engine) worker(ctx context.Context, id int, todo <-chan string) {
-	for chunk := range todo {
-		log.Debug("msg", "vacuuming a chunk", "worker", id, "chunk", chunk)
-		sql := fmt.Sprintf(sqlVacuumFmt, chunk)
-		numberVacuumConnections.Inc()
-		_, err := e.pool.Exec(ctx, sql)
-		numberVacuumConnections.Dec()
-		if err != nil {
-			log.Error("msg", "failed to vacuum chunk", "chunk", chunk, "worker", id, "error", err)
-			vacuumErrorsTotal.Inc()
-			// don't return error here. attempt to vacuum other chunks. keep working
-		} else {
-			tablesVacuumedTotal.Inc()
-		}
+func (e *Engine) worker(ctx context.Context, id int, todo <-chan *chunk) {
+	for c := range todo {
+		func() {
+			con, err := e.pool.Acquire(ctx)
```

Comment:
My thinking was that by getting a connection from the pool on each iteration, it would give other maintenance jobs an opportunity to run if we have fewer connections in the pool than maintenance activities.

But let's go with the simpler code and switch back if we need to.

Reply:
https://github.com/timescale/promscale/pull/1761/commits/d5103b5d05d168f087a85aeafa4f7463c4420e40